From bcac2add4aa433e3ca77117ccb6c11177b7b5b3a Mon Sep 17 00:00:00 2001
From: Gary Oliver <go@aerodesic.com>
Date: Wed, 6 Dec 2017 16:55:41 -0800
Subject: [PATCH] Change mutex_is_locked to return TRUE rather than FALSE

---
 fs/ubifs/ubifs.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/fs/ubifs/ubifs.h b/fs/ubifs/ubifs.h
index 1d89465..f85dde5 100644
--- a/fs/ubifs/ubifs.h
+++ b/fs/ubifs/ubifs.h
@@ -611,7 +611,7 @@ static inline ino_t parent_ino(struct dentry *dentry)
 /* misc.h */
 #define mutex_lock_nested(...)
 #define mutex_unlock_nested(...)
-#define mutex_is_locked(...)	0
+#define mutex_is_locked(...)	1  /* How did this ever work? Change to return TRUE instead of FALSE; Gary Oliver <go@aerodesic.com> */
 #endif
 
 /* Version of this UBIFS implementation */
-- 
2.7.4

